### PR TITLE
Fixes #25390 - pass post via template API

### PR DESCRIPTION
**Pull request by lzap**
This is needed for https://github.com/theforeman/community-templates/pull/461 to work via proxy

### CODE REVIEW

**Comment by timogoebel** [comment]
This needs a separate redmine issue.

File: lib/proxy/request.rb [hunk 1]
```diff
@@ -26,15 +26,16 @@ def uri(path)
...
         req.add_field(k, v)
       end
       req
     end
 
-    def create_post(path, body, headers={})
-      req = Net::HTTP::Post.new(uri(path).path)
+    def create_post(path, body, headers={}, query={})
+      uri = uri(path)
+      req = Net::HTTP::Post.new("#{uri.path || '/'}?#{query_string(query)}")
```
Comment (timogoebel) [inline lib/proxy/request.rb]: If there is a query string, can we set this in the uri? Manually building query string just causes trouble in my experience.

```
uri.query = query_string(query)
```

File: modules/templates/template_proxy_request.rb [hunk 2]
```diff
@@ -19,8 +19,16 @@ def get_template(kind, env, params)
       # in hostgroup provisioning there are spaces
       kind = kind.split('/').map{|x| CGI.escape(x)}.join('/')
       logger.debug "Template: request for #{kind} using #{opts.inspect} at #{uri.host}"
-      proxy_headers = extract_request_headers(env).merge("X-Forwarded-For" => "#{env['REMOTE_ADDR']}, #{proxy_ip}")
-      proxy_req = request_factory.create_get("/unattended/#{kind}", opts, proxy_headers)
+      proxy_headers = extract_request_headers(env)
+      proxy_headers.merge!("X-Forwarded-For" => "#{env['REMOTE_ADDR']}, #{proxy_ip}")
```
Comment (timogoebel) [inline modules/templates/template_proxy_request.rb]: You can avoid merge here,

```suggestion
      proxy_headers['X-Forwarded-For'] = "#{env['REMOTE_ADDR']}, #{proxy_ip}"
```

**Review by timogoebel — CHANGES_REQUESTED** [review]

File: lib/proxy/request.rb (hunk 1 above)
Comment (lzap) [reply lib/proxy/request.rb]: Sure.

File: modules/templates/template_proxy_request.rb (hunk 2 above)
Comment (lzap) [reply modules/templates/template_proxy_request.rb]: Copy pasted it, yeah :-)

**Comment by lzap** [comment]
Amended the changes + monkey patch for Hash.compact (this is Ruby 2.4+ only) to fix tests on 2.0-2.3. Created RM issue for this.

File: modules/templates/templates_api.rb [hunk 3]
```diff
@@ -25,4 +25,11 @@ class Proxy::TemplatesApi < Sinatra::Base
       Proxy::Templates::TemplateProxyRequest.new.get_template([kind].join('/'), request.env, params)
     end
   end
+
+  post "/:kind" do |kind|
+    log_halt(500, "Failed to retrieve #{kind} template for #{params.inspect}: ") do
+      params.merge!("Content-Type" => "text/plain")
```
Comment (timogoebel) [inline modules/templates/templates_api.rb]: I still see a `merge!` here. ;-)

Comment (lzap) [reply modules/templates/templates_api.rb]: Missed it somehow, fixed.

**Comment by lzap** [comment]
All checks are finally passing on all Ruby versions. This patch needs to be tested with new template change. The way I test this is just a fake host, template preview and then I launch wget/curl from my laptop to exit the build.

**Review by timogoebel — APPROVED** [review]
Code looks good to me. I don't have time to test this, sorry.

**Comment by lzap** [comment]
All relevant PRs merged, rebased, @iNecas interested in trying this out?

**Comment by iNecas** [comment]
I'm interested, I just need to get my devel provisioning setup working again. I wont get to it this week, adding to my TODO list for next one

File: modules/templates/templates_api.rb [hunk 4]
```diff
@@ -25,4 +25,11 @@ class Proxy::TemplatesApi < Sinatra::Base
       Proxy::Templates::TemplateProxyRequest.new.get_template([kind].join('/'), request.env, params)
     end
   end
+
+  post "/:kind" do |kind|
+    log_halt(500, "Failed to retrieve #{kind} template for #{params.inspect}: ") do
```
Comment (iNecas) [inline modules/templates/templates_api.rb]: Are we actually trying to retrieve template here? Perhaps more generic message would be a bit better

File: modules/templates/template_proxy_request.rb [hunk 5]
```diff
@@ -29,6 +37,14 @@ def get_template(kind, env, params)
       res.body
     end
 
+    def get_template(kind, env, params)
```
Comment (iNecas) [inline modules/templates/template_proxy_request.rb]: Since it's not about just tempaltes anymore, what about calling the methods just `get` and `post`?

**Comment by lzap** [comment]
All done

**Comment by lzap** [comment]
Rebased not sure why tests are not passing, locally fine.

**Review by iNecas — APPROVED** [review]
Tested and works well. Thanks @lzap

**Comment by lzap** [comment]
Thanks a buch to @iNecas for review and test!

**Pull request merged**